native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvCreator.java

```diff
@@ -0,0 +1,179 @@
+                                                  Field currentField) {
+        Object currentCsv = sm.currentCsvNode;
+        Object nilValue = config.nilValue;
+        if (sm.config.nilAsOptionalField && !type.isNilable()
+                && CsvUtils.isNullValue(nilValue, value)
+                && currentField != null && SymbolFlags.isFlagOn(currentField.getFlags(), SymbolFlags.OPTIONAL)) {
+            return null;
+        }
+        Object convertedValue = convertToExpectedType(StringUtils.fromString(value), type, config);
+        sm.isCurrentCsvNodeEmpty = false;
+        if (convertedValue instanceof BError || convertedValue instanceof CsvUtils.UnMappedValue) {
+            if (ignoreIncompatibilityErrorsForMaps(sm, exptype)) {
+                return null;
+            }
+            throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_CAST, value, type);
+        }
+
+        Type currentCsvNodeType = TypeUtils.getType(currentCsv);
+        switch (currentCsvNodeType.getTag()) {
+            case TypeTags.MAP_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
+                ((BMap<BString, Object>) currentCsv).put(StringUtils.fromString(getHeaderValueForColumnIndex(sm)),
+                        convertedValue);
+                return currentCsv;
+            case TypeTags.ARRAY_TAG:
+                ArrayType arrayType = (ArrayType) currentCsvNodeType;
+                if (arrayType.getState() == ArrayType.ArrayState.CLOSED &&
+                        arrayType.getSize() - 1 < sm.columnIndex) {
+                    return currentCsv;
+                }
+                ((BArray) currentCsv).add(sm.columnIndex, convertedValue);
+                return currentCsv;
+            case TypeTags.TUPLE_TAG:
+                ((BArray) currentCsv).add(sm.columnIndex, convertedValue);
+                return currentCsv;
+            default:
+                return convertedValue;
+        }
+    }
+
+    public static String getHeaderValueForColumnIndex(CsvParser.StateMachine sm) {
+        if (sm.config.customHeader == null &&  (sm.config.header == Boolean.FALSE)) {
+            String header = String.valueOf(sm.columnIndex + 1);
+            Map<String, Field> fieldHierarchy = sm.fieldHierarchy;
+            if (fieldHierarchy.containsKey(header)) {
+                fieldHierarchy.remove(header);
+            }
+            return header;
+        }
+        if (sm.columnIndex >= sm.headers.size()) {
+            throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_CUSTOM_HEADER_LENGTH);
+        }
+        String header = sm.headers.get(sm.columnIndex);
+        return getUpdatedHeaders(sm.updatedRecordFieldNames, header,
+                sm.fields.contains(header));
+    }
+
+    public static void checkAndAddCustomHeaders(CsvParser.StateMachine sm, Object customHeader) {
+        if (customHeader == null) {
+            return;
+        }
```

Comment:
Do we need to allow people to call `checkAndAddCustomHeaders` if they infact don't have a `customHeader`? From the name I feel like it's the responsibility of the call not to call this if they don't have it. WDYT?